src/dbnode/storage/series/buffer.go

```diff
@@ -183,78 +240,138 @@ func (b *dbBuffer) Write(
 	annotation []byte,
 ) error {
 	now := b.nowFn()
-	futureLimit := now.Add(1 * b.bufferFuture)
-	pastLimit := now.Add(-1 * b.bufferPast)
-	if !futureLimit.After(timestamp) {
-		return m3dberrors.ErrTooFuture
+	bucketStart := timestamp.Truncate(b.blockSize)
+	key := b.writableBucketKey(timestamp)
+
+	if b.isRealTime(timestamp) {
+		idx := b.writableBucketIdx(timestamp)
+
+		if b.bucketsRealTime[idx].needsReset(now, bucketStart) {
+			b.DrainAndReset()
+		}
+		// If there was previously a non-realtime metric written to this bucket
+		// already, then use that bucket.
+		if _, ok := b.bucketsNotRealTime[key]; ok {
+			b.makeBucketRealTime(key, idx)
+		}
+
+		return b.bucketsRealTime[idx].write(now, timestamp, value, unit, annotation)
 	}
-	if !pastLimit.Before(timestamp) {
-		return m3dberrors.ErrTooPast
+
+	if !b.opts.RetentionOptions().AnyWriteTimeEnabled() {
+		return errAnyWriteTimeNotEnabled
 	}
 
-	bucketStart := timestamp.Truncate(b.blockSize)
-	idx := b.writableBucketIdx(timestamp)
-	if b.buckets[idx].needsReset(bucketStart) {
-		// Needs reset
-		b.DrainAndReset()
+	if _, ok := b.bucketsNotRealTime[key]; !ok {
+		b.initializeBucket(key)
 	}
 
-	return b.buckets[idx].write(timestamp, value, unit, annotation)
+	if b.bucketsNotRealTime[key].needsDrain(now, bucketStart) {
+		b.bucketDrain(now, bucketIDNotRealtime(key), bucketStart)
+
+		if b.bucketsNotRealTime[key].isStale(now) {
```

Comment:
I was thinking to remove buckets from the map only when they haven't been written to for a while to prevent buckets from having to constantly get destroyed only to be recreated right away again.

At the same time, if there is a constant and sustained write to a bucket, I don't want there to be a large amount of metrics that never get drained. Therefore we want to drain if the bucket `.isFull()`, but we don't want to remove it yet because there's probably another metric going into this bucket soon.